Throw an error if the ASCII String containing the '+' symbol and ASCI…

qchaupds opened the pull request:
Throw error when field format contains '+' for ASCII String and '-' for ASCII Number fields
 
Closes NASA-PDS/validate#11
 
The issue is validate did not flag a bad ASCII_String format [%+8s].  The reference manual states that ASCII_String and a few other related should be left justified.
With the new code, it does now flag it.
The code also flag any ASCII number format if they contain the minus symbol as in [%-2d]

Tested in DEV:

validate -R pds4.label  -r report_github11_bad_1.json -s json -t src/test/resources/github11/test_data/science_index_bad_1.xml

validate -R pds4.label  -r report_github11_bad_2.json -s json -t src/test/resources/github11/test_data/science_index_bad_2.xml

validate -R pds4.label --skip-context-validation -r report_github11_good.json -s json -t src/test/resources/github11/test_data/science_index_good.xml

The first 2 tests should fail with 3 errors each: 2 for context and 1 for bad format.
The last test with the --skip-context-validation should validate successfully of the format.

If the last test was ran without the --skip-context-validation, there will be two errors for context.   It was left in since it may be confusing to have a failed test
for a valid input.

Commit:
Throw an error if the ASCII String containing the '+' symbol and ASCII number containing the '-' symbols.

1. Add test artifacts for github11
2. Add several new functions to validate the ASCII String and ASCII Number in TableDataContentValidationRule.java
3. Add github11 to ValidationIntegrationTests.java

Refs #11 https://github.com/nasa-pds/validate/issues/11 Update allowable field_format values per Standards Reference definition regarding [+|-] characters

Commit:
Improve performance by moving lists creation from functions to private variables in class declaration:

1. Move lists creation (ASCII_STRING_TYPE_LIST, and ASCII_NUMBER_TYPE_LIST) from functions to private variables TableDataContentValidationRule.java
2. Add error reporting of inconsistent number of fields in TableDataContentValidationRule.java
3. Comment out unused variable fieldDescriptionList in TableDataContentValidationRule.java
4. Comment out 'totalErrors' check for invalid tests in github11 to prevent from false negative ValidationIntegrationTests.java

Refs #11 https://github.com/nasa-pds/validate/issues/11 Update allowable field_format values per Standards Reference definition regarding [+|-] characters

Commit:
Improve logic to handle integration test input that does not have Product_Observational or Product_Observational/File_Area_Observational/Table_Character label:

1. Add logic for when a label does not have Product_Observational label or Product_Observational/File_Area_Observational/Table_Character
   since not all labels are required to contain those labels.
2. Add label name to error message to enhance debugging for developer.

Refs #11 https://github.com/nasa-pds/validate/issues/11 Update allowable field_format values per Standards Reference definition regarding [+|-] characters

Commit:
Remove extraneous class left over from development:

1. Remove unused class import

Refs #11 https://github.com/nasa-pds/validate/issues/11 Update allowable field_format values per Standards Reference definition regarding [+|-] characters

jordanpadams commented on src/main/java/gov/nasa/pds/tools/validate/rule/pds4/TableDataContentValidationRule.java:
```diff
@@ -84,9 +94,40 @@
+
+  private static String PRODUCT_OBSERVATIONAL = "Product_Observational";
+  private static String TABLE_CHARACTER = "Product_Observational/File_Area_Observational/Table_Character";
+  private static String RECORD_CHARACTER = "Product_Observational/File_Area_Observational/Table_Character/Record_Character";
```
@qchaupds could we use a shorter absolute XPath here instead of including this nesting? that way we do not have to update the code whenever a new product is introduced that uses it. e.g.

```
//Table_Character
//Table_Character/Record_Character
```

thoughts?

jordanpadams commented on src/main/java/gov/nasa/pds/tools/validate/rule/pds4/TableDataContentValidationRule.java:
```diff
@@ -133,11 +174,136 @@ public boolean isApplicable(String location) {
+    for (String oneFieldFormat : fieldFormatList) {
+        //LOG.debug("validateAsciiStringFieldsFormat:indexToList,oneFieldFormat,fieldTypeList.get(indexToList)  {},{},{}",indexToList,oneFieldFormat,fieldTypeList.get(indexToList));
+        // Check for prohibited plus symbol '+' in format field : %+8s
+        if (this.ASCII_STRING_TYPE_LIST.contains(fieldTypeList.get(indexToList).toUpperCase())) {
```
`ASCII_STRING_TYPE_LIST` is a static variable, so we shouldn't be using `this.` here.

jordanpadams commented on src/main/java/gov/nasa/pds/tools/validate/rule/pds4/TableDataContentValidationRule.java:
```diff
@@ -133,11 +174,136 @@ public boolean isApplicable(String location) {
+    for (String oneFieldFormat : fieldFormatList) {
+        //LOG.debug("validateAsciiNumberFieldsFormat:indexToList,oneFieldFormat,fieldTypeList.get(indexToList)  {},{},{}",indexToList,oneFieldFormat,fieldTypeList.get(indexToList));
+        // Check for prohibited minus symbol '-' in format field : %-2d
+        if (this.ASCII_NUMBER_TYPE_LIST.contains(fieldTypeList.get(indexToList).toUpperCase())) {
```
`ASCII_NUMBER_TYPE_LIST` is a static variable, so we shouldn't be using `this.` here.

jordanpadams commented on src/main/java/gov/nasa/pds/tools/validate/rule/pds4/TableDataContentValidationRule.java:
```diff
@@ -133,11 +174,136 @@ public boolean isApplicable(String location) {
+
+     try {
+        XMLExtractor extractor = new XMLExtractor(getTarget());
+        TinyNodeImpl productObservationalNode = extractor.getNodeFromDoc(this.PRODUCT_OBSERVATIONAL);
```
looks like static variables are being accessed with `this` throughout. can we just do a quick scrub of this new code to make sure we are accessing this properly.

jordanpadams commented on src/main/java/gov/nasa/pds/tools/validate/rule/pds4/TableDataContentValidationRule.java:
```diff
@@ -133,11 +174,136 @@ public boolean isApplicable(String location) {
+            if (oneFieldFormat.contains("%-")) {
+                LOG.error("ASCII Number related fields should not contain '-' symbol in field_format [" + oneFieldFormat + "] in field_number " + fieldNumberList.get(indexToList));
+                addTableProblem(ExceptionType.ERROR, 
+                    ProblemType.TABLE_FILE_READ_ERROR,
```
I think we should probably use the `ProblemType.INVALID_LABEL` since this is really a bad label specification, not the data itself. does that seem reasonable?

jordanpadams reviewed (changes requested):
@qchaupds functionality looks great and works as expected. commented with a few things that need to updated.

also, I think this code improvement should go in the `TableFieldDefinitionRule` class since what we are really catching here is errors in the definition of table fields, not the content of the data itself. I could be convinced otherwise, but just a thought.

thanks for the great work!

Commit:
Move validation of a table format in a label from TableDataContentValidationRule to TableFieldDefinitionRule:

1. Remove validation code from TableDataContentValidationRule.java
2. Remove unused imports statements from TableDataContentValidationRule.java
3. Add validation code to TableFieldDefinitionRule.java
4. Add TableFieldDefinitionRule to validation-commands.xml so TableFieldDefinitionRule can be executed on a label
5. Update problem type from TABLE_FILE_READ_ERROR to INVALID_LABEL when reporting error to match reported error from TableFieldDefinitionRule.java in ValidationIntegrationTests.java

Refs #11 https://github.com/nasa-pds/validate/issues/11 Update allowable field_format values per Standards Reference definition regarding [+|-] characters

jordanpadams committed:
Remove erroneous files

jordanpadams committed:
Update test data to remove/update invalid context refs

jordanpadams reviewed (approved):
@qchaupds looks great!

note: i updated the test data to update/remove those invalid context references so we no longer have those erroneous other errors to worry about.
